File: src/lib/matcher.ts
import { GitHubLabel, MatchResult } from './types';

// lowercases and trims label name
export function normalizeLabel(name: string): string {
  return (name || '').trim().toLowerCase();
}

/**
 * Checks if issue labels match tracked labels.
 * @param issueLabels Labels on the issue
 * @param trackedLabels Labels being tracked
 * @param matchMode 'any' (at least 1) or 'all' (all required)
 */
export function matchIssueLabels(
  issueLabels: (GitHubLabel | string)[],
  trackedLabels: string[],
  matchMode: 'any' | 'all' = 'any'
): MatchResult {
  if (!trackedLabels || trackedLabels.length === 0) {
    return {
      isMatch: false,
      matchedLabels: [],
      totalTracked: 0,
      totalMatched: 0,
      matchMode,
      reason: 'No tracked labels configured for this repository',
    };
  }

  // track all mode
  if (trackedLabels.includes('__ALL__')) {
    const allLabelNames = issueLabels.map((l) =>
      typeof l === 'string' ? l : l.name
    );
    return {
      isMatch: true,
      matchedLabels: allLabelNames,
      totalTracked: trackedLabels.length,
      totalMatched: allLabelNames.length,
      matchMode,
      reason: 'Tracking all issues from this repository',
    };
  }

  const issueLabelNames = issueLabels.map((l) =>
    typeof l === 'string' ? normalizeLabel(l) : normalizeLabel(l.name)
  );

  const matchedLabels: string[] = [];

  for (const tracked of trackedLabels) {
    const normTracked = normalizeLabel(tracked);
    const foundIndex = issueLabelNames.indexOf(normTracked);
    if (foundIndex !== -1) {
      // preserve original casing
      const original = typeof issueLabels[foundIndex] === 'string'
        ? (issueLabels[foundIndex] as string)
        : (issueLabels[foundIndex] as GitHubLabel).name;
      matchedLabels.push(original);
    }
  }

  const isMatch =
    matchMode === 'all'
      ? matchedLabels.length === trackedLabels.length
      : matchedLabels.length > 0;

  let reason = '';
  if (isMatch) {
    reason =
      matchMode === 'all'
        ? `Matched all ${trackedLabels.length} tracked labels: ${matchedLabels.join(', ')}`
        : `Matched ${matchedLabels.length} of ${trackedLabels.length} tracked labels: ${matchedLabels.join(', ')}`;
  } else {
    reason =
      matchMode === 'all'
        ? `Missing some required labels (found ${matchedLabels.length}/${trackedLabels.length})`
        : `None of the issue labels matched any of the ${trackedLabels.length} tracked labels`;
  }

  return {
    isMatch,
    matchedLabels,
    totalTracked: trackedLabels.length,
    totalMatched: matchedLabels.length,
    matchMode,
    reason,
  };
}

File: app/core/matcher.py
"""
Label matching engine.
"""

from __future__ import annotations

from typing import Union


def normalize_label(name: str) -> str:
    """Lowercase and strip a label name."""
    return (name or "").strip().lower()


def match_issue_labels(
    issue_labels: list[Union[dict, str]],
    tracked_labels: list[str],
    match_mode: str = "any",
) -> dict:
    """Check whether an issue's labels match the tracked set."""
    if not tracked_labels:
        return {
            "isMatch": False,
            "matchedLabels": [],
            "totalTracked": 0,
            "totalMatched": 0,
            "matchMode": match_mode,
            "reason": "No tracked labels configured for this repository",
        }

    # "__ALL__" means track every issue regardless of labels
    if "__ALL__" in tracked_labels:
        all_names = [
            (l if isinstance(l, str) else l.get("name", "")) for l in issue_labels
        ]
        return {
            "isMatch": True,
            "matchedLabels": all_names,
            "totalTracked": len(tracked_labels),
            "totalMatched": len(all_names),
            "matchMode": match_mode,
            "reason": "Tracking all issues from this repository",
        }

    issue_label_names = [
        normalize_label(l) if isinstance(l, str) else normalize_label(l.get("name", ""))
        for l in issue_labels
    ]

    matched_labels: list[str] = []
    for tracked in tracked_labels:
        norm = normalize_label(tracked)
        if norm in issue_label_names:
            idx = issue_label_names.index(norm)
            original = (
                issue_labels[idx]
                if isinstance(issue_labels[idx], str)
                else issue_labels[idx].get("name", "")
            )
            matched_labels.append(original)

    if match_mode == "all":
        is_match = len(matched_labels) == len(tracked_labels)
    else:
        is_match = len(matched_labels) > 0

    if is_match:
        reason = (
            f"Matched all {len(tracked_labels)} tracked labels: {', '.join(matched_labels)}"
            if match_mode == "all"
            else f"Matched {len(matched_labels)} of {len(tracked_labels)} tracked labels: {', '.join(matched_labels)}"
        )
    else:
        reason = (
            f"Missing some required labels (found {len(matched_labels)}/{len(tracked_labels)})"
            if match_mode == "all"
            else f"None of the issue labels matched any of the {len(tracked_labels)} tracked labels"
        )

    return {
        "isMatch": is_match,
        "matchedLabels": matched_labels,
        "totalTracked": len(tracked_labels),
        "totalMatched": len(matched_labels),
        "matchMode": match_mode,
        "reason": reason,
    }
